src/main/java/org/peergos/client/NabuClient.java

```diff
@@ -0,0 +1,265 @@
+package org.peergos.client;
+
+import io.ipfs.cid.Cid;
+import io.ipfs.multiaddr.MultiAddress;
+import io.ipfs.multihash.Multihash;
+import org.peergos.util.JSONParser;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.ConnectException;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.LinkedHashMap;
+import java.util.stream.Collectors;
+
+public class NabuClient {
+
+    private static final int DEFAULT_CONNECT_TIMEOUT_MILLIS = 10_000;
+    private static final int DEFAULT_READ_TIMEOUT_MILLIS = 60_000;
+
+    public final String host;
+    public final int port;
+    public final String protocol;
+    private final String apiVersion;
+    private final int connectTimeoutMillis;
+    private final int readTimeoutMillis;
+
+    public NabuClient(String host, int port) {
+        this(host, port, "/api/v0/", false);
+    }
+
+    public NabuClient(String multiaddr) {
+        this(new MultiAddress(multiaddr));
+    }
+
+    public NabuClient(MultiAddress addr) {
+        this(addr.getHost(), addr.getPort(), "/api/v0/", detectSSL(addr));
+    }
+
+    public NabuClient(String host, int port, String version, boolean ssl) {
+        this(host, port, version, DEFAULT_CONNECT_TIMEOUT_MILLIS, DEFAULT_READ_TIMEOUT_MILLIS, ssl);
+    }
+
+    public NabuClient(String host, int port, String version, int connectTimeoutMillis, int readTimeoutMillis, boolean ssl) {
+        if (connectTimeoutMillis < 0) throw new IllegalArgumentException("connect timeout must be zero or positive");
+        if (readTimeoutMillis < 0) throw new IllegalArgumentException("read timeout must be zero or positive");
+        this.host = host;
+        this.port = port;
+        this.connectTimeoutMillis = connectTimeoutMillis;
+        this.readTimeoutMillis = readTimeoutMillis;
+
+        if (ssl) {
+            this.protocol = "https";
+        } else {
+            this.protocol = "http";
+        }
+
+        this.apiVersion = version;
+    }
+
+    public NabuClient timeout(int timeout) {
+        return new NabuClient(host, port, apiVersion, timeout, timeout, protocol.equals("https"));
+    }
+
+    public String version() throws IOException {
+        Map m = (Map)retrieveAndParse("version");
+        return (String)m.get("Version");
+    }
+
+    public Map id() throws IOException {
```

Comment:
Can we make this return the peerid as a cid?

Reply:
done